# Task 4.9: Ibra Lisso

---------------------------------

# Part 1 Task



<b>1.1. Download the customer data set and add it to your “Original Data” folder.

<b>1.2. Create a new notebook in your “Scripts” folder for part 1 of this task.
    
    Directives Completed

# 01. Importing Libraries

<b>Import your analysis libraries, as well as your new customer data set as a dataframe.

In [2]:
#Importing Libraries;

import numpy as np
import pandas as pd
import os

In [3]:
#Creating a Folderpath; 

path = '/users/ibu/Documents/12-2022 Instacart Basket Analysis'

In [4]:
#Importing the 'Customers Dataset' from 'Original Data' Folder

Customers_Dataset = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers_data.csv'))

In [6]:
#Checking Data Columns

Customers_Dataset.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

<b>Notes:</b> 

#Updating all headers into 'small caps'

#Fixing columns [n-dependants = total_dependants & fam_status = marital_status]

In [7]:
#Checking dataset shape

Customers_Dataset.shape

(206209, 10)

In [26]:
#Performing Descriptive Statitics to Get Clues; 

Customers_Dataset.describe()

user_id            age  total_dependents         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

<b>Notes:</b> 
1. Oldest Customer is 81 years, Youngest is 18 years.
2. We some Customers have no dependents, most have more than 1 dependent
3. We can see no Customer earns more than 594,000 or lower than 25,000

# 02. Data Wrangling

In [9]:
#Fixing Column Headers by renaming each of them; 

Customers_Dataset.rename(columns = {'n_dependants': 'total_dependents'}, inplace = True)

Customers_Dataset.rename(columns = {'fam_status': 'marital_status'}, inplace = True)

In [12]:
#Fixing Column-Headers by Changing from 'CAPS' to 'caps'; 

Customers_Dataset.rename(columns = {'STATE': 'state'}, inplace = True)
Customers_Dataset.rename(columns = {'Surnam': 'last_name'}, inplace = True)
Customers_Dataset.rename(columns = {'First Name': 'first_name'}, inplace = True)
Customers_Dataset.rename(columns = {'Age': 'age'}, inplace = True)
Customers_Dataset.rename(columns = {'Gender': 'gender'}, inplace = True)

In [13]:
#Checking Columns Counts and Shape; 

Customers_Dataset.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   total_dependents marital_status  income  
0                 3        married  165665  
1                 0         single   59285  
2                 2        married   99568  
3                 0         single   42049  
4                 1        married   40374

In [14]:
Customers_Dataset.shape

(206209, 10)

<b>Notes:</b> Columns and Mispelled-Text Fixes completed and Customers Dataset Counts and Shape matches.

# 03. Performing Data Quality & Consistency Checks 

   <b>Guided Steps:</b> Missing Values<b>,</b> Duplicates & Mixed-Type Data.

In [15]:
#Finding Missing Data Values;

Customers_Dataset.isnull().sum()

user_id                 0
first_name          11259
last_name               0
gender                  0
state                   0
age                     0
date_joined             0
total_dependents        0
marital_status          0
income                  0
dtype: int64

<b>Notes:</b> Seems there are 11,259 missing First Names. That is equivalent to 11,259 rows in total. 

In [16]:
#Checking the Missing Data values on 'first_name';

df_nan = Customers_Dataset[Customers_Dataset['first_name'].isnull() == True]

In [17]:
df_nan

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        total_dependents marital_status  income  
53                     2        married   41709  
73                     0         single   82518  
82                     3        married  117099  
99                     1        married  155673  
105                    3        married  151819  
...                  ...            ...     ...  
206038                 3        married   87783  
206044                 2        married  108488  
206090                 1        married   45275  
206162                 0         single   56325  
206171                 2        married   59222  

[11259 rows x 10 columns]

<b>Notes:</b> Yes, indeed 11,259 Customer's First Names are missing

In [18]:
#Performing Data Types Check of all Columns; 

Customers_Dataset.dtypes

user_id              int64
first_name          object
last_name           object
gender              object
state               object
age                  int64
date_joined         object
total_dependents     int64
marital_status      object
income               int64
dtype: object

In [19]:
#Checking on Mixed-Data Types; 

for col in Customers_Dataset.columns.tolist():
  weird = (Customers_Dataset[[col]].applymap(type) != Customers_Dataset[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (Customers_Dataset[weird]) > 0:
    print (col)

first_name


In [20]:
#Converting Data-Type 'First Name' to String;

Customers_Dataset['first_name'] = Customers_Dataset['first_name'].astype('str')

In [21]:
#Re-Checking for Mixed-Types after Converting 'First Name' into String;

for col in Customers_Dataset.columns.tolist():
  weird = (Customers_Dataset[[col]].applymap(type) != Customers_Dataset[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (Customers_Dataset[weird]) > 0:
    print (col)

<b>Notes:</b> No further Mixed-Types Data exists after conversion to strings of existing Mixed-Type data.

In [22]:
#Checking for any Data Duplicates; 

Customers_Dataset[Customers_Dataset.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, total_dependents, marital_status, income]
Index: []

<b>Notes:</b> No duplicates exists. Dataset good on this part.

# 04. Combine the Datasets: 'Customers' & 'Instacart Orders_Products';

<b>Key Hint:</b> Ensure the key columns matches (same data type)

In [23]:
#Importing the 'Ords_Prods_T48' Dataset' from 'Prepared Data' Folder;

df_Ords_Prods_T48 = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'Ords_Prods_T48.pkl'))

In [27]:
#Checking 'Ords_Prods_T48' Dataset Columns and Rows; 

df_Ords_Prods_T48.head(5)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...   prices_range_loc     Busiest day    Busiest_days  \
0          0  ...  Mid Range Product  Regularly busy  Regularly busy   
1          1  ...  Mid Range Product  Regularly busy    Slowest days   
2          1  ...  Mid Range Product  Regularly busy    Slowest days   
3          1  ...  Mid Range Product      Least busy    Slowest days   
4          1  ...  Mid Range Product      Least busy    Slowest days   

   Busiest_period_of_day  max_order   loyalty_flag Average_Prices  \
0         Average Orders         10  New Customers           14.0   
1         Average Orders         10  New Customers           14.0   
2            Most Orders         10  New Customers           14.0   
3         Average Orders         10  New Customers           14.0   
4            Most Orders         10  New Customers           14.0   

    spender_flag median_days-since_prior_order          frequency_flag  
0  High Spenders                          20.5  Non-Frequent Customers  
1  High Spenders                          20.5  Non-Frequent Customers  
2  High Spenders                          20.5  Non-Frequent Customers  
3  High Spenders                          20.5  Non-Frequent Customers  
4  High Spenders                          20.5  Non-Frequent Customers  

[5 rows x 26 columns]

In [28]:
#Re-Checking 'Customers_Dataset' Columns and Rows for common key 'column name'; 

Customers_Dataset.head(5)

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   total_dependents marital_status  income  
0                 3        married  165665  
1                 0         single   59285  
2                 2        married   99568  
3                 0         single   42049  
4                 1        married   40374

<b>Notes:</b> The key column we see between the two datasets: 'Customers_Dataset & Ords_Prods_T48 = <b>user_id</b>

In [31]:
#Checking the shape of 'df_Ords_Prods_T48' Dataset; 

df_Ords_Prods_T48.shape

(32404859, 26)

<b>Notes:</b> 

1. <b>Customers_Dataset</b> has 206,209 Rows + 10 Columns
2. <b>df_Ords_Prods_T48</b> has 32,404,859 Rows + 26 Columns

In [32]:
#Carrying Out the Merging of the Two Datasets; 

Cust_Ords_Prods_T49 = df_Ords_Prods_T48.merge(Customers_Dataset, on = 'user_id')

In [33]:
#Checking the shape of 'Cust_Ords_Prods_T49' Dataset; 

Cust_Ords_Prods_T49.shape

(32404859, 35)

<b>Notes:</b> The new combined Cust_Ords_Prods dataset now has same 32,404,859 rows number, but more columns = 35

In [34]:
#Checking the Columns + Rows of the New 'Cust_Ords_Prods' Dataset;

Cust_Ords_Prods_T49.head(10)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   
5             5   3367565        1             6                   2   
6             6    550135        1             7                   1   
7             7   3108588        1             8                   1   
8             8   2295261        1             9                   1   
9             9   2550362        1            10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered  ...          frequency_flag first_name  last_name  gender  \
0          0  ...  Non-Frequent Customers      Linda     Nguyen  Female   
1          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   
2          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   
3          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   
4          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   
5          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   
6          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   
7          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   
8          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   
9          1  ...  Non-Frequent Customers      Linda     Nguyen  Female   

     state age date_joined total_dependents marital_status income  
0  Alabama  31   2/17/2019                3        married  40423  
1  Alabama  31   2/17/2019                3        married  40423  
2  Alabama  31   2/17/2019                3        married  40423  
3  Alabama  31   2/17/2019                3        married  40423  
4  Alabama  31   2/17/2019                3        married  40423  
5  Alabama  31   2/17/2019                3        married  40423  
6  Alabama  31   2/17/2019                3        married  40423  
7  Alabama  31   2/17/2019                3        married  40423  
8  Alabama  31   2/17/2019                3        married  40423  
9  Alabama  31   2/17/2019                3        married  40423  

[10 rows x 35 columns]

# 05. Exporting to 'Prepared Data' Folder as 'pkl' 

<b>Notes:</b> Shape and Columns Checked for verification purposes. Now time to export it for Part 2 of Task 4.9.

In [35]:
Cust_Ords_Prods_T49.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'Cust_Ords_Prods_T49A.pkl'))

In [36]:
Cust_Ords_Prods_T49.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Cust_Ords_Prods_T49A.csv'))